File: js/kpiEngine.js
/**
 * KPI Engine Module
 * Computes all business metrics, aggregations, trend tables, and forecasting logic.
 */

window.KpiEngine = (function () {

    /**
     * Filter the datasets based on active controls
     */
    function filterData(data, filters) {
        if (!data) return null;

        let sales = [...data.sales];
        let refunds = [...data.refunds];
        let targets = [...data.targets];
        let batches = [...data.batches];

        if (filters.category && filters.category !== 'All') {
            sales = sales.filter(s => s.Category === filters.category);
            refunds = refunds.filter(r => r.Category === filters.category);
            targets = targets.filter(t => t.Category === filters.category);
        }

        if (filters.batch && filters.batch !== 'All') {
            sales = sales.filter(s => s.Batch === filters.batch);
            refunds = refunds.filter(r => r.Batch === filters.batch);
            batches = batches.filter(b => b.Batch === filters.batch);
        }

        if (filters.teacher && filters.teacher !== 'All') {
            sales = sales.filter(s => s.Teacher === filters.teacher);
            refunds = refunds.filter(r => r.Teacher === filters.teacher);
            batches = batches.filter(b => b.Teacher === filters.teacher);
        }

        if (filters.state && filters.state !== 'All') {
            sales = sales.filter(s => s.State === filters.state);
            refunds = refunds.filter(r => r.State === filters.state);
        }

        return { sales, refunds, targets, batches };
    }

    /**
     * Calculate core KPIs for the filtered dataset
     */
    function calculateKPIs(filteredData) {
        const { sales, refunds, targets, batches } = filteredData;

        // Base metrics
        const totalGrossRevenue = sales.reduce((acc, s) => acc + s.Sales, 0);
        const totalRefundAmount = refunds.reduce((acc, r) => acc + r.RefundAmount, 0);
        const netRevenue = Math.max(0, totalGrossRevenue - totalRefundAmount);
        const refundPercent = totalGrossRevenue > 0 ? (totalRefundAmount / totalGrossRevenue) * 100 : 0;
        const totalOrders = sales.length;
        const averageSellingPrice = totalOrders > 0 ? totalGrossRevenue / totalOrders : 0;

        // Calculate Today's & Yesterday's Sales
        let todaySales = 0;
        let yesterdaySales = 0;
        const uniqueDates = [...new Set(sales.map(s => s.Date))].sort();

        if (uniqueDates.length > 0) {
            const maxDateStr = uniqueDates[uniqueDates.length - 1];
            todaySales = sales.filter(s => s.Date === maxDateStr).reduce((acc, s) => acc + s.Sales, 0);

            if (uniqueDates.length > 1) {
                const prevDateStr = uniqueDates[uniqueDates.length - 2];
                yesterdaySales = sales.filter(s => s.Date === prevDateStr).reduce((acc, s) => acc + s.Sales, 0);
            }
        }

        // Calculate exact calendar span of the dataset
        let daysInPeriod = 1;
        if (uniqueDates.length > 0) {
            const minDate = new Date(uniqueDates[0]);
            const maxDate = new Date(uniqueDates[uniqueDates.length - 1]);
            const diffTime = Math.abs(maxDate - minDate);
            const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24)) + 1;
            daysInPeriod = isNaN(diffDays) ? uniqueDates.length : diffDays;
        }

        // Daily Run Rate (DRR) over calendar span
        const dailyRunRate = totalGrossRevenue / (daysInPeriod || 1);

        // Growth metrics with safety checks
        let wowGrowth = 0;
        if (uniqueDates.length >= 14) {
            const last7Dates = uniqueDates.slice(-7);
            const prev7Dates = uniqueDates.slice(-14, -7);

            const last7Sales = sales.filter(s => last7Dates.includes(s.Date)).reduce((acc, s) => acc + s.Sales, 0);
            const prev7Sales = sales.filter(s => prev7Dates.includes(s.Date)).reduce((acc, s) => acc + s.Sales, 0);

            wowGrowth = prev7Sales > 0 ? ((last7Sales - prev7Sales) / prev7Sales) * 100 : 0;
        }

        let momGrowth = 0;
        if (uniqueDates.length >= 2) {
            const mid = Math.floor(uniqueDates.length / 2);
            const secondHalfDates = uniqueDates.slice(mid);
            const firstHalfDates = uniqueDates.slice(0, mid);

            const secondHalfSales = sales.filter(s => secondHalfDates.includes(s.Date)).reduce((acc, s) => acc + s.Sales, 0);
            const firstHalfSales = sales.filter(s => firstHalfDates.includes(s.Date)).reduce((acc, s) => acc + s.Sales, 0);

            momGrowth = firstHalfSales > 0 ? ((secondHalfSales - firstHalfSales) / firstHalfSales) * 100 : 0;
        }

        // Target progress calculations
        const totalTarget = targets.reduce((acc, t) => acc + t.Target, 0);
        const targetAchievedPercent = totalTarget > 0 ? (totalGrossRevenue / totalTarget) * 100 : 0;
        const targetRemaining = Math.max(0, totalTarget - totalGrossRevenue);

        // Days in month calculation dynamically determined from the last transaction
        let daysInMonth = 30;
        let daysPassed = 1;
        let daysRemaining = 29;
        if (uniqueDates.length > 0) {
            const lastDate = new Date(uniqueDates[uniqueDates.length - 1]);
            const year = lastDate.getFullYear();
            const month = lastDate.getMonth(); // 0-indexed
            daysInMonth = new Date(year, month + 1, 0).getDate(); // exact calendar days in month
            daysPassed = lastDate.getDate();
            daysRemaining = Math.max(1, daysInMonth - daysPassed);
        }

        const requiredDRR = targetRemaining / daysRemaining;

        // Forecast metrics based on dynamic month-end
        const expectedMonthRevenue = dailyRunRate * daysInMonth;
        const forecastTargetAchievement = totalTarget > 0 ? (expectedMonthRevenue / totalTarget) * 100 : 0;

        // Profit & Loss calculation (from batches or estimated)
        const profit = batches.reduce((acc, b) => acc + (b.Profit || 0), 0) || (netRevenue * 0.45);
        const loss = totalRefundAmount;

        return {
            grossRevenue: totalGrossRevenue,
            netRevenue,
            refundAmount: totalRefundAmount,
            refundPercent,
            orders: totalOrders,
            averageSellingPrice,
            todaySales,
            yesterdaySales,
            dailyRunRate,
            wowGrowth,
            momGrowth,
            totalTarget,
            targetAchievedPercent,
            targetRemaining,
            requiredDRR,
            expectedMonthRevenue,
            forecastTargetAchievement,
            profit,
            loss
        };
    }

    /**
     * Entity specific analysis generators
     */
    function getCategoryPerformance(filteredData) {
        const { sales } = filteredData;
        const categories = [...new Set(sales.map(s => s.Category))];
        return categories.map(cat => {
            const catSales = sales.filter(s => s.Category === cat);
            const revenue = catSales.reduce((acc, s) => acc + s.Sales, 0);
            const orders = catSales.length;
            const profit = parseFloat((revenue * 0.45).toFixed(2));
            return { category: cat, revenue, orders, profit };
        }).sort((a, b) => b.revenue - a.revenue);
    }

    function getBatchPerformance(filteredData) {
        const { sales, refunds, batches } = filteredData;
        const batchNames = [...new Set(sales.map(s => s.Batch))];

        return batchNames.map(bName => {
            const batchSales = sales.filter(s => s.Batch === bName);
            const batchRefunds = refunds.filter(r => r.Batch === bName);
            const revenue = batchSales.reduce((acc, s) => acc + s.Sales, 0);
            const refundAmount = batchRefunds.reduce((acc, r) => acc + r.RefundAmount, 0);
            const admissions = batchSales.length - batchRefunds.length;

            const batchObj = batches.find(b => b.Batch === bName) || { Capacity: 50, Profit: revenue * 0.45 };
            const fillPercent = batchObj.Capacity > 0 ? (admissions / batchObj.Capacity) * 100 : 0;

            return {
                batch: bName,
                revenue,
                admissions,
                capacity: batchObj.Capacity,
                fillPercent,
                profit: batchObj.Profit || (revenue - refundAmount) * 0.45,
                refundAmount
            };
        }).sort((a, b) => b.revenue - a.revenue);
    }

    function getTeacherPerformance(filteredData) {
        const { sales } = filteredData;
        const teachers = [...new Set(sales.map(s => s.Teacher))];
        return teachers.map(t => {
            const teacherSales = sales.filter(s => s.Teacher === t);
            const revenue = teacherSales.reduce((acc, s) => acc + s.Sales, 0);
            const admissions = teacherSales.length;
            return { teacher: t, revenue, admissions };
        }).sort((a, b) => b.revenue - a.revenue);
    }

    function getStatePerformance(filteredData) {
        const { sales } = filteredData;
        const states = [...new Set(sales.map(s => s.State))];
        return states.map(st => {
            const stateSales = sales.filter(s => s.State === st);
            const revenue = stateSales.reduce((acc, s) => acc + s.Sales, 0);
            return { state: st, revenue };
        }).sort((a, b) => b.revenue - a.revenue);
    }

    /**
     * Compute comparative metrics for two specific entities (e.g. Batch vs Batch)
     */
    function compareBatches(data, batchA, batchB) {
        const dataA = filterData(data, { batch: batchA });
        const dataB = filterData(data, { batch: batchB });

        const kpiA = calculateKPIs(dataA);
        const kpiB = calculateKPIs(dataB);

        const perfA = getBatchPerformance(dataA).find(b => b.batch === batchA) || {};
        const perfB = getBatchPerformance(dataB).find(b => b.batch === batchB) || {};

        const metrics = [
            { label: 'Sales Orders', valA: kpiA.orders, valB: kpiB.orders, format: 'num' },
            { label: 'Revenue', valA: kpiA.grossRevenue, valB: kpiB.grossRevenue, format: 'currency' },
            { label: 'Growth % (MoM)', valA: kpiA.momGrowth, valB: kpiB.momGrowth, format: 'percent' },
            { label: 'Refund %', valA: kpiA.refundPercent, valB: kpiB.refundPercent, format: 'percent' },
            { label: 'Profit', valA: kpiA.profit, valB: kpiB.profit, format: 'currency' },
            { label: 'Target Achievement %', valA: kpiA.targetAchievedPercent, valB: kpiB.targetAchievedPercent, format: 'percent' },
            { label: 'Batch Fill %', valA: perfA.fillPercent || 0, valB: perfB.fillPercent || 0, format: 'percent' }
        ];

        return metrics.map(m => {
            const diff = m.valA - m.valB;
            const pctDiff = m.valB > 0 ? (diff / m.valB) * 100 : 0;
            return {
                metric: m.label,
                valA: m.valA,
                valB: m.valB,
                diff,
                pctDiff,
                format: m.format
            };
        });
    }

    return {
        filterData,
        calculateKPIs,
        getCategoryPerformance,
        getBatchPerformance,
        getTeacherPerformance,
        getStatePerformance,
        compareBatches
    };
})();

File: backend/app/services/kpi_engine.py
from datetime import date, datetime, timedelta
import calendar
from typing import Dict, List, Optional, Any
import pandas as pd

# Each entry: {"margin": float, "is_default": bool}
# is_default=True means this is a placeholder estimate, not a verified figure.
DEFAULT_CATEGORY_MARGINS = {
    "Coding & Tech":     {"margin": 0.50, "is_default": False},
    "K-12 Academics":    {"margin": 0.40, "is_default": False},
    "Creative Arts":     {"margin": 0.45, "is_default": True},
    "Language Learning": {"margin": 0.35, "is_default": True},
    "Default":           {"margin": 0.45, "is_default": True},
}

def calculate_kpis(
    sales_df: pd.DataFrame,
    refunds_df: Optional[pd.DataFrame] = None,
    targets_df: Optional[pd.DataFrame] = None,
    batches_df: Optional[pd.DataFrame] = None,
    category_margins: Optional[Dict[str, Dict]] = None,
    reference_date: Optional[date] = None
) -> Dict[str, Any]:
    """
    Computes all core business KPIs, entity breakdowns, and forecasts.
    Fixes key bugs:
    1. Today & Yesterday use real calendar system dates (or injectable reference_date).
    2. MoM growth uses an equal-window comparison: days 1–days_passed of current month vs
       days 1–days_passed of the previous month. Output includes `days_compared` (the window size).
    3. Profit margins are configurable per category via {"margin": float, "is_default": bool} dicts.
    4. Missing target data returns `target_data_missing: true` (no fabricated target).
    5. Missing batch capacity flags `capacity_missing: true` (no default to 50).
    """
    if category_margins is None:
        category_margins = DEFAULT_CATEGORY_MARGINS

    ref_date = reference_date if reference_date else date.today()
    yesterday_date = ref_date - timedelta(days=1)

    # Standardize Sales DataFrame
    sales = sales_df.copy() if (sales_df is not None and not sales_df.empty) else pd.DataFrame(columns=["Date", "Sales", "Category", "Batch", "Teacher", "State", "OrderID"])
    if "Date" in sales.columns and not sales.empty:
        sales["Date_parsed"] = pd.to_datetime(sales["Date"], errors="coerce").dt.date
    else:
        sales["Date_parsed"] = pd.Series(dtype="object")

    if "Sales" in sales.columns:
        sales["Sales"] = pd.to_numeric(sales["Sales"], errors="coerce").fillna(0.0)
    else:
        sales["Sales"] = 0.0

    # Standardize Refunds DataFrame
    refunds = refunds_df.copy() if (refunds_df is not None and not refunds_df.empty) else pd.DataFrame(columns=["Date", "RefundAmount", "Category", "Batch", "Teacher", "State"])
    if "RefundAmount" in refunds.columns:
        refunds["RefundAmount"] = pd.to_numeric(refunds["RefundAmount"], errors="coerce").fillna(0.0)
    else:
        refunds["RefundAmount"] = 0.0

    # 1. Base Revenue & Order Metrics
    gross_revenue = float(sales["Sales"].sum()) if not sales.empty else 0.0
    refund_amount = float(refunds["RefundAmount"].sum()) if not refunds.empty else 0.0
    net_revenue = max(0.0, gross_revenue - refund_amount)
    refund_percent = float((refund_amount / gross_revenue) * 100.0) if gross_revenue > 0 else 0.0
    total_orders = len(sales)
    average_selling_price = float(gross_revenue / total_orders) if total_orders > 0 else 0.0

    # 2. System Calendar Date Today & Yesterday Sales
    today_sales = float(sales[sales["Date_parsed"] == ref_date]["Sales"].sum()) if not sales.empty else 0.0
    yesterday_sales = float(sales[sales["Date_parsed"] == yesterday_date]["Sales"].sum()) if not sales.empty else 0.0

    # 3. Daily Run Rate (DRR) over dataset's actual date span
    valid_dates = sales["Date_parsed"].dropna().unique()
    if len(valid_dates) > 0:
        min_d = min(valid_dates)
        max_d = max(valid_dates)
        days_in_period = max(1, (max_d - min_d).days + 1)
    else:
        days_in_period = 1
    daily_run_rate = float(gross_revenue / days_in_period)

    # 4. WoW Growth (last 7 days vs previous 7 days relative to ref_date or latest date)
    wow_anchor = ref_date if ref_date in valid_dates else (max(valid_dates) if len(valid_dates) > 0 else ref_date)
    last_7_start = wow_anchor - timedelta(days=6)
    prev_7_start = wow_anchor - timedelta(days=13)
    prev_7_end = wow_anchor - timedelta(days=7)

    last_7_sales = float(sales[(sales["Date_parsed"] >= last_7_start) & (sales["Date_parsed"] <= wow_anchor)]["Sales"].sum()) if not sales.empty else 0.0
    prev_7_sales = float(sales[(sales["Date_parsed"] >= prev_7_start) & (sales["Date_parsed"] <= prev_7_end)]["Sales"].sum()) if not sales.empty else 0.0

    wow_growth = float(((last_7_sales - prev_7_sales) / prev_7_sales) * 100.0) if prev_7_sales > 0 else 0.0

    # 5. MoM Growth — equal-window comparison
    # Compare days 1..days_passed of the current month vs days 1..days_passed of the previous month.
    # e.g. if today is Aug 10, compare Aug 1-10 vs Jul 1-10 (not the full July).
    current_year = ref_date.year
    current_month = ref_date.month
    days_passed = ref_date.day  # e.g. 10 on Aug 10

    if current_month == 1:
        prev_month = 12
        prev_month_year = current_year - 1
    else:
        prev_month = current_month - 1
        prev_month_year = current_year

    def is_in_month_window(d_val, year, month, day_limit):
        """True if date is within days 1..day_limit of the given month/year."""
        return (
            d_val is not None
            and d_val.year == year
            and d_val.month == month
            and d_val.day <= day_limit
        )

    current_month_sales = float(
        sales[sales["Date_parsed"].apply(
            lambda d: is_in_month_window(d, current_year, current_month, days_passed)
        )]["Sales"].sum()
    ) if not sales.empty else 0.0

    prev_month_sales = float(
        sales[sales["Date_parsed"].apply(
            lambda d: is_in_month_window(d, prev_month_year, prev_month, days_passed)
        )]["Sales"].sum()
    ) if not sales.empty else 0.0

    mom_growth = float(((current_month_sales - prev_month_sales) / prev_month_sales) * 100.0) if prev_month_sales > 0 else 0.0
    days_compared = days_passed  # window size for both months

    # 6. Targets Data & Forecasting
    target_data_missing = (targets_df is None or targets_df.empty or "Target" not in targets_df.columns)
    
    if target_data_missing:
        total_target = None
        target_achieved_percent = None
        target_remaining = None
        required_drr = None
        expected_month_revenue = None
        forecast_target_achievement = None
    else:
        targets_df_clean = targets_df.copy()
        targets_df_clean["Target"] = pd.to_numeric(targets_df_clean["Target"], errors="coerce").fillna(0.0)
        total_target = float(targets_df_clean["Target"].sum())

        target_achieved_percent = float((gross_revenue / total_target) * 100.0) if total_target > 0 else 0.0
        target_remaining = float(max(0.0, total_target - gross_revenue))

        # Days in month based on ref_date
        _, days_in_month = calendar.monthrange(ref_date.year, ref_date.month)
        days_passed = ref_date.day
        days_remaining = max(1, days_in_month - days_passed)

        required_drr = float(target_remaining / days_remaining)
        expected_month_revenue = float(daily_run_rate * days_in_month)
        forecast_target_achievement = float((expected_month_revenue / total_target) * 100.0) if total_target > 0 else 0.0

    # 7. Category Performance & Configurable Profit Margins
    _fallback_margin_info = {"margin": 0.45, "is_default": True}

    def _get_margin_info(cat: str) -> Dict:
        """Resolve margin info dict for a category, falling back to Default then hardcoded."""
        return (
            category_margins.get(cat)
            or category_margins.get("Default")
            or _fallback_margin_info
        )

    category_perf = []
    total_profit = 0.0
    if not sales.empty and "Category" in sales.columns:
        for cat_name, group in sales.groupby("Category"):
            cat_str = str(cat_name)
            rev = float(group["Sales"].sum())
            orders = len(group)
            m_info = _get_margin_info(cat_str)
            margin = m_info["margin"]
            cat_profit = float(round(rev * margin, 2))
            total_profit += cat_profit
            category_perf.append({
                "category": cat_str,
                "revenue": round(rev, 2),
                "orders": orders,
                "profit": cat_profit,
                "margin_used": margin,
                "is_default": m_info.get("is_default", True),
            })
        category_perf.sort(key=lambda x: x["revenue"], reverse=True)

    # 8. Batch Performance & Missing Capacity Flagging
    batch_perf = []
    if not sales.empty and "Batch" in sales.columns:
        batches_map = {}
        if batches_df is not None and not batches_df.empty and "Batch" in batches_df.columns:
            for _, b_row in batches_df.iterrows():
                b_name = str(b_row["Batch"])
                b_cap = b_row.get("Capacity") if "Capacity" in batches_df.columns else None
                b_profit = b_row.get("Profit") if "Profit" in batches_df.columns else None
                batches_map[b_name] = {"capacity": b_cap, "profit": b_profit}

        for b_name, group in sales.groupby("Batch"):
            b_str = str(b_name)
            b_rev = float(group["Sales"].sum())
            b_orders = len(group)
            
            # refunds for this batch
            b_refunds = refunds[refunds["Batch"] == b_str] if (not refunds.empty and "Batch" in refunds.columns) else pd.DataFrame()
            b_refund_amt = float(b_refunds["RefundAmount"].sum()) if not b_refunds.empty else 0.0
            net_admissions = b_orders - len(b_refunds)

            batch_info = batches_map.get(b_str, {})
            cap_val = batch_info.get("capacity")

            capacity_missing = (cap_val is None or pd.isna(cap_val))
            if capacity_missing:
                capacity = None
                fill_percent = None
            else:
                try:
                    capacity = int(cap_val)
                    fill_percent = float((net_admissions / capacity) * 100.0) if capacity > 0 else 0.0
                except (ValueError, TypeError):
                    capacity = None
                    fill_percent = None
                    capacity_missing = True

            cat_str = str(group["Category"].iloc[0]) if "Category" in group.columns else "Default"
            b_m_info = _get_margin_info(cat_str)
            b_margin = b_m_info["margin"]
            b_profit = float(batch_info.get("profit")) if (batch_info.get("profit") is not None and not pd.isna(batch_info.get("profit"))) else float(round((b_rev - b_refund_amt) * b_margin, 2))

            batch_perf.append({
                "batch": b_str,
                "revenue": round(b_rev, 2),
                "admissions": net_admissions,
                "capacity": capacity,
                "fill_percent": round(fill_percent, 2) if fill_percent is not None else None,
                "capacity_missing": capacity_missing,
                "profit": b_profit,
                "refund_amount": round(b_refund_amt, 2)
            })
        batch_perf.sort(key=lambda x: x["revenue"], reverse=True)

    # 9. Teacher Performance
    teacher_perf = []
    if not sales.empty and "Teacher" in sales.columns:
        for t_name, group in sales.groupby("Teacher"):
            t_rev = float(group["Sales"].sum())
            t_adm = len(group)
            teacher_perf.append({
                "teacher": str(t_name),
                "revenue": round(t_rev, 2),
                "admissions": t_adm
            })
        teacher_perf.sort(key=lambda x: x["revenue"], reverse=True)

    # 10. State Performance
    state_perf = []
    if not sales.empty and "State" in sales.columns:
        for st_name, group in sales.groupby("State"):
            st_rev = float(group["Sales"].sum())
            state_perf.append({
                "state": str(st_name),
                "revenue": round(st_rev, 2)
            })
        state_perf.sort(key=lambda x: x["revenue"], reverse=True)

    loss = refund_amount

    return {
        "gross_revenue": round(gross_revenue, 2),
        "net_revenue": round(net_revenue, 2),
        "refund_amount": round(refund_amount, 2),
        "refund_percent": round(refund_percent, 2),
        "orders": total_orders,
        "average_selling_price": round(average_selling_price, 2),
        "today_sales": round(today_sales, 2),
        "yesterday_sales": round(yesterday_sales, 2),
        "daily_run_rate": round(daily_run_rate, 2),
        "wow_growth": round(wow_growth, 2),
        "mom_growth": round(mom_growth, 2),
        "days_compared": days_compared,
        "target_data_missing": target_data_missing,
        "total_target": round(total_target, 2) if total_target is not None else None,
        "target_achieved_percent": round(target_achieved_percent, 2) if target_achieved_percent is not None else None,
        "target_remaining": round(target_remaining, 2) if target_remaining is not None else None,
        "required_drr": round(required_drr, 2) if required_drr is not None else None,
        "expected_month_revenue": round(expected_month_revenue, 2) if expected_month_revenue is not None else None,
        "forecast_target_achievement": round(forecast_target_achievement, 2) if forecast_target_achievement is not None else None,
        "profit": round(total_profit, 2),
        "loss": round(loss, 2),
        "category_performance": category_perf,
        "batch_performance": batch_perf,
        "teacher_performance": teacher_perf,
        "state_performance": state_perf
    }
